eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -87,108 +115,87 @@ object IncomingPaymentPacket {
+            }
+          case None if add.blinding_opt.isDefined => Left(InvalidOnionPayload(UInt64(12), 0))
+          case None =>
             // We check if the payment is using trampoline: if it is, we may not be the final recipient.
-            payload.records.get[OnionPaymentPayloadTlv.TrampolineOnion] match {
+            payload.get[OnionPaymentPayloadTlv.TrampolineOnion] match {
               case Some(OnionPaymentPayloadTlv.TrampolineOnion(trampolinePacket)) =>
-                // NB: when we enable blinded trampoline routes, we will need to check if the outer onion contains a blinding
-                // point and use it to derive the decryption key for the blinded trampoline onion.
-                decryptOnion(add.paymentHash, privateKey, trampolinePacket, PaymentOnionCodecs.trampolineOnionPerHopPayloadCodec) match {
-                  case Left(failure) => Left(failure)
-                  case Right(DecodedOnionPacket(innerPayload: PaymentOnion.NodeRelayPayload, next)) => validateNodeRelay(add, payload, innerPayload, next)
-                  case Right(DecodedOnionPacket(innerPayload: PaymentOnion.FinalTlvPayload, _)) => validateFinal(add, payload, innerPayload)
-                  case Right(DecodedOnionPacket(_: PaymentOnion.BlindedFinalPayload, _)) => Left(InvalidOnionPayload(UInt64(12), 0)) // trampoline blinded routes are not supported yet
+                // NB: when we enable blinded trampoline routes, we will need to check if the outer onion contains a
+                // blinding point and use it to derive the decryption key for the blinded trampoline onion.
+                decryptOnion(add.paymentHash, privateKey, trampolinePacket).flatMap {
+                  case DecodedOnionPacket(innerPayload, Some(next)) => validateNodeRelay(add, payload, innerPayload, next)
+                  case DecodedOnionPacket(innerPayload, None) => validateFinalPayload(add, payload, innerPayload)
                 }
-              case None => validateFinal(add, payload)
+              case None => validateFinalPayload(add, payload)
             }
-          case _: PaymentOnion.BlindedFinalPayload =>
-            // TODO: receiving through blinded routes is not supported yet.
-            Left(InvalidOnionPayload(UInt64(12), 0))
         }
     }
   }
 
-  private def isValidBlindedPayment(data: BlindedRouteData.PaymentData, amount: MilliSatoshi, cltvExpiry: CltvExpiry, features: Features[Feature]): Boolean = {
-    val amountOk = amount >= data.paymentConstraints.minAmount
-    val expiryOk = cltvExpiry <= data.paymentConstraints.maxCltvExpiry
-    val featuresOk = Features.areCompatible(features, data.allowedFeatures)
-    amountOk && expiryOk && featuresOk
+  private def validateBlindedChannelRelayPayload(add: UpdateAddHtlc, payload: TlvStream[OnionPaymentPayloadTlv], blindedPayload: TlvStream[RouteBlindingEncryptedDataTlv], nextBlinding: PublicKey, nextPacket: OnionRoutingPacket): Either[FailureMessage, IncomingPaymentPacket] = {
```

Comment:
Same on other methods:
```suggestion
  private def validateBlindedChannelRelayPayload(add: UpdateAddHtlc, payload: TlvStream[OnionPaymentPayloadTlv], blindedPayload: TlvStream[RouteBlindingEncryptedDataTlv], nextBlinding: PublicKey, nextPacket: OnionRoutingPacket): Either[FailureMessage, ChannelRelayPacket] = {
```

I wonder if, you could have used types instead of methods for the validation. E.g.: `TentativeChannelRelayHopPayload`/`ChannelRelayHopPayload`? Probably overkill and useless.

Reply:
Fixed in https://github.com/ACINQ/eclair/pull/2414/commits/318e59e425ebe19877db1800523aa31e788995f8